serveradmin/nessus/models.py

```diff
@@ -0,0 +1,299 @@
+            'Referer': self.base,
+            'X-Requested-With': 'XMLHttpRequest',
+            'Connection': 'keep-alive',
+            'X-Api-Token': self.api_token
+        }
+
+        self.sess_token = self.login()
+
+        if len(self.sess_token) > 5:
+            self.session.headers['X-Cookie'] = "token=" + self.sess_token
+        elif self.access_key and self.secret_key:
+            self.logger.info('Using API keys')
+            self.api_keys = True
+            self.session.headers['X-ApiKeys'] = 'accessKey=%s; secretKey=%s' % (self.access_key, self.secret_key)
+
+
+    def login(self):
+        """
+        Log user into nessus.
+
+        return string auth token that needs to be send with every later request in a cookie X-Cookie: token=<token>.
+        """
+        data = {"username": self.user, "password": self.password}
+        response = self.request('/session', method='POST', data=json.dumps(data))
+        if "Invalid Credentials" in response.text:
+            self.logger.error('Invalid credentials provided! Cannot authenticate to Nessus.')
+            raise Exception('[FAIL] Invalid credentials provided! Cannot authenticate to Nessus.')
+        if response.status_code != 200:
+            self.logger.error('Couldn\'t authenticate! Error returned by Nessus: %s' % (json.loads(response.text)['error']))
+            raise Exception('[FAIL] Couldn\'t authenticate! Error returned by Nessus: %s' % (json.loads(response.text)['error']))
+        self.logger.info('Logged in to Nessus using password authentication and X-Api-Token - %s' % (self.api_token))
+        return json.loads(response.text)['token']
+
+    def get_api_token(self) -> None:
+        """Refresh X-Api-Token value."""
+        response = self.request('/nessus6.js?v=1642551183681', method='get')
+        offset = response.text.index('return g(a,[{key:"getApiToken",value:function(){')
+        token = re.findall(r'return"(.*?)"\}\}', response.text[offset:offset + 100])
+
+        if not token[0]:
+            self.logger.error('Could not get new X-Api-Token from Nessus')
+            raise Exception('Could not get new X-Api-Token from Nessus')
+        self.api_token = token[0]
+        self.session.headers['X-Api-Token'] = self.api_token
+        self.logger.info('Got new X-Api-Token from Nessus - %s' % (self.api_token))
+
+    def request(self, url, data=None, method='POST', download=False, json_output=False):
+        """
+        Send request to nessus.
+
+        :data: request body to send.
+        :method: request method GET, POST, PUT, DELETE.
+        :downnload: True for downloading scan results in csv format.
+        :json_output: True / False.
+
+        return dict
+        """
+        timeout = 0
+        success = False
+        method = method.lower()
+        url = self.base + url
+        self.logger.info('Requesting to url %s' % (url))
+
+        while (timeout <= 30) and (not success):
```

Comment:
It's a bit hard to put this function here, because different http status codes from nessus need to trigger different responses. So if it's okay with you, I would leave this in. If you still want me to switch to retries, I will see what I can do,